### ISSUE TITLE
gluon.sqlhtml: give possibility to make options_widget if not field.requires

### ISSUE THREAD

**Issue by sbourgeois1** [web2py/web2py#1418]
Suggestion:
    def widget(cls, field, value, **attributes):
        """
        Generates a SELECT tag, including OPTIONs (only 1 option allowed)

```
    see also: `FormWidget.widget`
    """
    default = dict(value=value)
    attr = cls._attributes(field, default,
                           **attributes)
    if '_options' in attributes and isinstance(attributes['_options'], (list, tuple)):# suggestion
        options = attributes['_options'] # suggestion
    else:
        requires = field.requires
        if not isinstance(requires, (list, tuple)):
            requires = [requires]
        if requires:
            if hasattr(requires[0], 'options'):
                options = requires[0].options()
        else:
            raise SyntaxError(
                'widget cannot determine options of %s' % (field))
    opts = [OPTION(v, _value=k) for (k, v) in options]
    return SELECT(*opts, **attr)
```

**Comment by niphlod** [comment]
I'm quite lost at trying to understand what this accomplishes.... can you explain ?

**Comment by niphlod** [comment]
@sbourgeois1 : if you can't provide a real-life example, I'm going to close this issue in a few days.

**Comment by sbourgeois1** [comment]
Hello again,
Thanks for tackeling this problem.

I tried to transfer you a zip directory containing the application, but the file is too big for the transfer to work.
As the problem occured a while ago, I need some time to find out the function which failed and prepare a test.I hope being back this afternoon. Très cordialement,Serge Bourgeois
Mob: +32 477 33 10 11
Mail: serge.bourgeois@yahoo.fr

 
     
----- Mail transféré -----
 De : Serge Bourgeois <serge.bourgeois@yahoo.fr>
 À : web2py/web2py <reply@reply.github.com> 
 Envoyé le : Jeudi 22 septembre 2016 10h55
 Objet : Re: [web2py/web2py] gluon.sqlhtml: give possibility to make options_widget if not field.requires (#1418)
   
Hi,Sorry for not responding immediately...

You'll find here attached a zip directory containing an application where I was obliged to find a turn aound...I'll come back to you soon with more details 
 Très cordialement,Serge Bourgeois
Mob: +32 477 33 10 11
Mail: serge.bourgeois@yahoo.fr

 
      De : niphlod <notifications@github.com>
 À : web2py/web2py <web2py@noreply.github.com> 
Cc : sbourgeois1 <serge.bourgeois@yahoo.fr>; Mention <mention@noreply.github.com>
 Envoyé le : Mercredi 21 septembre 2016 22h53
 Objet : Re: [web2py/web2py] gluon.sqlhtml: give possibility to make options_widget if not field.requires (#1418)
  
@sbourgeois1 : if you can't provide a real-life example, I'm going to close this issue in a few days.—

**Comment by sbourgeois1** [comment]
Hello,
The problem occurs when it is necessary to create a tuple of options using a function. Then the options are transfered to the view using  '_options' .

I hope the example will help understanding the problem. 

Extract of the db the model: 

table and fields directly involved :
...db.define_table('t_bks_trans_entry',# financial transactions
 ...
     Field('f_ac', 'reference t_bks_ac', requires = IS_EMPTY_OR(IS_IN_DB(db((db.t_bks_ac.f_company == session.bks_company_id) & (db.t_bks_ac.f_usage == 2)), 't_bks_ac.id', lambda r: '%s %s' % (r.f_ac, r.f_ac_name), orderby = 'f_orderby, f_ac_grp, id')),  represent = lambda val, r = None: ('%s %s' % (db.t_bks_ac[val].f_ac, db.t_bks_ac[val].f_ac_name)) if val else None , label = T('Account'), comment = T('Required if no party')),....

      Field('f_party', 'reference t_bks_party', requires = IS_EMPTY_OR(IS_IN_DB(db((db.t_bks_party.f_company == session.bks_company_id) & (db.t_bks_party.is_active == 'T')), 't_bks_party.id', '%(f_party)s', multiple = False, orderby = 'f_party')),  represent = lambda val, r = None: db.t_bks_party[val].f_party if val else None , label = T('Party'), comment = T('Required if no account')),
...

In the controller/ function in charge of showing / creating / editing financial transactions:


... my_trans_entry =db.t_bks_trans_entry
 my_trans_entry.f_ac.represent = lambda val, row:options_widget_150px(my_trans_entry.f_ac, val, **{'_name':'f_ac_row_%s' % row.id, '_disabled': 'disabled'}) \ if not(db.t_bks_ac_grp[db.t_bks_ac[row.f_ac].f_ac_grp].f_third_party =='F') else \ options_widget_200px(my_trans_entry.f_ac, val,**{'_name': 'f_ac_row_%s' % row.id, '_options' : give_ac_options(val,row)})

 my_trans_entry.f_party.represent = lambda val,row: options_widget(my_trans_entry.f_party, val,**{'_name':'f_party_row_%s' % row.id, '_options' :give_party_options(val, row)}) \ if(db.t_bks_ac_grp[db.t_bks_ac[row.f_ac].f_ac_grp].f_third_party =='T') else \ options_widget(my_trans_entry.f_party, val,**{'_name':'f_party_row_%s' % row.id, '_disabled': 'disabled'})
…
In the same controller, some functions are in charge of preparing the tuple of valid options depending of the context.See here under

def give_ac_options(val, row): ''' returns a list of tuples used in options widget fortrans_entry_ac''' my_return = [] if 'f_party' in row and row.f_party: mysql = ' select id, f_ac, f_ac_name from t_bks_ac ' mysql += ' where t_bks_ac.id in (select ' if row.f_column == 'c': mysql += ' f_credit_ac ' else: mysql += ' f_debit_ac ' mysql += ' from t_bks_party where id = ' +str(row.f_party) + ' limit 1)' mysql += ' and t_bks_ac.is_active = \'T\' limit 1;' rows = db.executesql(mysql) if rows: for rw in rows: my_return.append(('%s' % (rw[0]), '%s %s' %(rw[1], rw[2]))) elif 'f_ac_grp' in row and row.f_ac_grp: mysql = ' select id, f_ac, f_ac_name from t_bks_ac ' mysql += ' where t_bks_ac.f_ac_grp = ' + str(row.f_ac_grp) my_ac_l = convert_list(row.f_l_ac) if len(my_ac_l): mysql += ' and id in (0' for my_ac in my_ac_l: mysql = mysql + ', ' + str(my_ac) mysql += ')' mysql += ' and t_bks_ac.is_active = \'T\' order byf_orderby, f_ac;' rows = db.executesql(mysql) if rows: for rw in rows: my_return.append(('%s' % (rw[0]), '%s %s' %(rw[1], rw[2]))) return my_returndef give_party_options(val, row): ''' returns a list of tuples used in options widget fortrans_entry_ac''' my_return = [] if 'f_ac_grp' in row and row.f_ac_grp: mysql = ' select t_bks_party.id, t_bks_party.f_party ' mysql += ' from t_bks_ac_grp ' mysql += ' join t_bks_party on t_bks_ac_grp.f_party_type =t_bks_party.f_party_type ' mysql += ' where t_bks_ac_grp.id = ' + str(row.f_ac_grp) my_party_l = convert_list(row.f_l_party) if len(my_party_l): mysql += ' and id in (0' for my_party in my_party_l: mysql = mysql + ', ' + str(my_party) mysql += ')' mysql += ' and t_bks_party.is_active = \'T\' order byf_party, id;' rows = db.executesql(mysql) if rows: for rw in rows: my_return.append(('%s' % (rw[0]), '%s' % (rw[1]))) return my_return Très cordialement,Serge Bourgeois
Mob: +32 477 33 10 11
Mail: serge.bourgeois@yahoo.fr

 
      De : niphlod <notifications@github.com>
 À : web2py/web2py <web2py@noreply.github.com> 
Cc : sbourgeois1 <serge.bourgeois@yahoo.fr>; Mention <mention@noreply.github.com>
 Envoyé le : Mercredi 21 septembre 2016 22h53
 Objet : Re: [web2py/web2py] gluon.sqlhtml: give possibility to make options_widget if not field.requires (#1418)
   
@sbourgeois1 : if you can't provide a real-life example, I'm going to close this issue in a few days.—

**Comment by niphlod** [comment]
@sbourgeois1 : wayyy to much information. Pack a MINIMAL app that reproduces and explains the issue, please.

**Comment by sbourgeois1** [comment]
Hello,
Here attached a test application that you can unzip and test with and without the change proposal. 
Once you enter in the url  'default/test/t_bks_company/t_bks_trans_header.f_company/1/t_bks_trans_entry.f_trans_header/1', pay attention to the droplist of the field 'Party': without the change, the droplist is not filtered, while it is with the proposed change. Thanks for devoting time for the WEB2PY community!

Serge
Mob: +32 477 33 10 11
Mail: serge.bourgeois@yahoo.fr

 
     

 
      De : niphlod <notifications@github.com>
 À : web2py/web2py <web2py@noreply.github.com> 
Cc : sbourgeois1 <serge.bourgeois@yahoo.fr>; Mention <mention@noreply.github.com>
 Envoyé le : Mercredi 21 septembre 2016 22h53
 Objet : Re: [web2py/web2py] gluon.sqlhtml: give possibility to make options_widget if not field.requires (#1418)
  
@sbourgeois1 : if you can't provide a real-life example, I'm going to close this issue in a few days.—

**Comment by niphlod** [comment]
@sbourgeois1 : please do mind that you should ost the attachment somewhere. it wan't attached to the issue at https://github.com/web2py/web2py/issues/1418

tl;dr: don't use reply-to to this. Go to github.com ! ^_^

**Comment by sbourgeois1** [comment]
If the attached file is not available, you can also download a zip file containing the application from my dropbox:
https://www.dropbox.com/s/ej1jcurmqjg586z/tst_bks.zip?dl=0
Place into your web2py folder the unzipped content of the attached tst_bks.zip file.
Serge
[tst_bks.zip](https://github.com/web2py/web2py/files/504343/tst_bks.zip)

**Comment by niphlod** [comment]
ok, I downloaded the app but it's way too much complicated for a minimal thing that shows the culprit. I'm trying my best here but still can't understand where web2py is falling short.
And you're constructing queries concatening strings...the logic is wayy to much convoluted to even try to get what you ask. 
Can't you just reduce to the bare minimum and explain once again WHY do you need something and WHAT FOR (and, also, what "that something" is in the first place)?

**Comment by sbourgeois1** [comment]
The problem occurs in smartgrid, when you have a referenced field, and when you want to control the droplist of this field via a lambda function, depending on the row fields. With the change proposal, you can fill the variable '_options' to control the droplist.
Très cordialement,Serge Bourgeois
Mob: +32 477 33 10 11
Mail: serge.bourgeois@yahoo.fr

 
      De : niphlod <notifications@github.com>
 À : web2py/web2py <web2py@noreply.github.com> 
Cc : sbourgeois1 <serge.bourgeois@yahoo.fr>; State change <state_change@noreply.github.com>
 Envoyé le : Samedi 1 octobre 2016 20h42
 Objet : Re: [web2py/web2py] gluon.sqlhtml: give possibility to make options_widget if not field.requires (#1418)
   
ok, I downloaded the app but it's way too much complicated for a minimal thing that shows the culprit. I'm trying my best here but still can't understand where web2py is falling short.
And you're constructing queries concatening strings...the logic is wayy to much convoluted to even try to get what you ask. 
Can't you just reduce to the bare minimum and explain once again WHY do you need something and WHAT FOR (and, also, what "that something" is in the first place)?—
